src/main/host/memory_manager.rs

```diff
@@ -0,0 +1,769 @@
+use super::syscall_types::*;
+use super::thread::{CThread, Thread};
+use crate::cbindings as c;
+use crate::utility::interval_map::{Interval, IntervalMap, Mutation};
+use crate::utility::proc_maps;
+use crate::utility::proc_maps::{MappingPath, Sharing};
+use std::cell::{Cell, RefCell};
+use std::collections::HashMap;
+use std::fs::File;
+use std::fs::OpenOptions;
+use std::os::raw::c_void;
+use std::os::unix::io::AsRawFd;
+use std::process;
+
+static HEAP_PROT: i32 = libc::PROT_READ | libc::PROT_WRITE;
+static STACK_PROT: i32 = libc::PROT_READ | libc::PROT_WRITE;
+
+fn page_size() -> usize {
+    unsafe { libc::sysconf(libc::_SC_PAGESIZE) as usize }
+}
+
+/// TODO: why doesn't `cargo test` run doc tests?
+/// ```
+/// let ps = page_size();
+/// assert_eq!(page_of(0), 0);
+/// assert_eq!(page_of(page_size()-1), 0);
+/// assert_eq!(page_of(page_size()), page_size());
+/// assert_eq!(page_of(2*page_size()), 2*page_size());
+/// assert_eq!(page_of(2*page_size())-1, page_size());
+/// ```
+fn page_of(p: usize) -> usize {
+    p & !(page_size() - 1)
+}
+
+// Represents a region of plugin memory.
+#[derive(Clone, Debug)]
+struct Region {
+    // Where the region is mapped into shadow's address space, or NULL if it isn't.
+    shadow_base: *mut c_void,
+    read: bool,
+    write: bool,
+    execute: bool,
+    sharing: proc_maps::Sharing,
+    // The *original* path. Not the path to our mem file.
+    original_path: Option<proc_maps::MappingPath>,
+}
+
+/// Manages the address-space for a plugin process.
+pub struct MemoryManager {
+    shm_file: RefCell<ShmFile>,
+    regions: RefCell<IntervalMap<Region>>,
+    misses_by_path: RefCell<HashMap<String, u32>>,
+
+    // We need to reinitialize some state after an exec, but need to wait until the next syscall.
+    need_post_exec_cleanup: Cell<bool>,
+
+    // The part of the stack that we've already remapped in the plugin.
+    // We initially mmap enough *address space* in Shadow to accomodate a large stack, but we only
+    // lazily allocate it. This is both to prevent wasting memory, and to handle that we can't map
+    // the current "working area" of the stack in thread-preload.
+    stack_copied: RefCell<Interval>,
+
+    // The bounds of the heap. Note that before the plugin's first `brk` syscall this will be a
+    // zero-sized interval (though in the case of thread-preload that'll have already happened
+    // before we get control).
+    heap: RefCell<Interval>,
+}
```

Comment:
I don't understand the reason for all of the RefCells here (except maybe `misses_by_path`). Why is interior mutability needed, rather than just taking the entire memory manager as mutable? Rust's ownership checks seem like they would be extra useful for an "unsafe" memory manager.

Reply:
If the methods that return immutable references took a mutable reference to the MemoryManager, then we'd never be able to have multiple outstanding immutable refs. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5b2c92c605755367ee7bbdd4f31e6dd1

Added some comments.

> Rust's ownership checks seem like they would be extra useful for an "unsafe" memory manager.

Yeah, that reminds me I'd had an idea to make this safer. I'd written some TODOs that dropped somewhere along the way.

Ultimately I'd like to keep track of the memory ranges of borrowed references using another IntervalMap, and enforce that mutable references are exclusive from any other borrowed references. I think to do that these'll need to return a Ref-like object that updates the MemoryManager's bookkeeping when they're dropped.

I'd rather defer that change for now, but I added some comments about it, and for now made the methods that return mutable references take `&mut self`.

Reply:
Thinking about this some more - maybe only allowing one outstanding ref at a time isn't actually that bad. Small values (e.g. integers) can just be copied out, and even if we end up doing multiple lookups of the same ref in some scenarios, that's probably cheaper than the bookkeeping I was thinking of where we have to mutate another intervalmap everytime a ref is borrowed or dropped.

Going to take a quick stab at removing internal mutability and just requiring a mutable ref in most (all?) of MemoryManager's methods, and see how it looks

Reply:
I tried it, and was almost ready to go with that approach, but realized that if we end up having to do multiple accesses of pointers that we're *not* mapping; i.e. that we have to use the thread->get[Readable|Writeable]Pointer APIs, it could be a significant performance hit. I think I'd rather leave it as-is for now.

I updated the comments again though to note that while we might consider tracking borrowed memory ranges, it might be better to just see if we can avoid needing to hold a mutable ref simultaneously with other refs. I suspect we largely can.

Anyway, the net result of this comment thread (so far) are some more detailed comments, and changing the methods that return a mutable ref to take `&mut self`, which I think are both improvements at least :)

Reply:
> If the methods that return immutable references took a mutable reference to the MemoryManager, then we'd never be able to have multiple outstanding immutable refs. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5b2c92c605755367ee7bbdd4f31e6dd1

Ah, makes sense.

Regarding the comment "Perhaps we could require the caller to explicitly call another hook on the first syscall after the `execve` is done", it seems like this would allow you to get rid of a few RefCells (`shm_file` and `regions`), so I don't think making the call explicit is a bad thing. But I'm okay with how it is now, thanks for adding the comments.

> Ultimately I'd like to keep track of the memory ranges of borrowed references using another IntervalMap, and enforce that mutable references are exclusive from any other borrowed references. I think to do that these'll need to return a Ref-like object that updates the MemoryManager's bookkeeping when they're dropped.
>
> I'd rather defer that change for now, but I added some comments about it, and for now made the methods that return mutable references take &mut self.

I think this is a good idea (but for a possible future PR). Another option (if willing to change the interface) could be a function or macro that allows you to take multiple mutable references at once so that the bookkeeping overhead isn't needed. Something like `(ref1, ref2, ref3) = manager_get_mut!(mem_manager, plugin_ptr_1, plugin_ptr_2, plugin_ptr_3)`, which checks that the memory doesn't overlap, but also keeps a mutable reference to the memory manager so that additional mutable references can't be taken.

> I tried it, and was almost ready to go with that approach, but realized that if we end up having to do multiple accesses of pointers that we're not mapping; i.e. that we have to use the thread->get[Readable|Writeable]Pointer APIs, it could be a significant performance hit. I think I'd rather leave it as-is for now.

Sounds good to me.

Reply:
> Something like (ref1, ref2, ref3) = manager_get_mut!(mem_manager, plugin_ptr_1, plugin_ptr_2, plugin_ptr_3), which checks that the memory doesn't overlap, but also keeps a mutable reference to the memory manager so that additional mutable references can't be taken.

Yeah, something like that might end up working.

Thinking about this some more, I've flip-flopped again and removed the internal mutability. I think preferring simplicity and static checking makes sense until we know for sure we need something else.

Reply:
Sorry for making you second-guess yourself :) Either approach is good with me.